criu/kerndat.c

```diff
@@ -1260,13 +1323,44 @@ static int kerndat_has_nftables_concat(void)
 #endif
 }
 
+/*
+ * Some features depend on resource that can be dynamically changed
+ * at the OS runtime. There are cases that we cannot determine the
+ * availability of those features at the first time we run kerndat
+ * check. So in later kerndat checks, we need to retry to get those
+ * information. This function contains calls to those kerndat checks.
+ *
+ * Those kerndat checks must
+ * Return -1 on error
+ * Return 0 when the check is successful but no new information
+ * Return 1 when the check is successful and there is new information
+ */
+int kerndat_try_load_new(void)
+{
+	int ret, has_new_data = 0;
+
+	ret = kerndat_get_hugetlb_dev();
+	if (ret < 0)
+		return ret;
+	has_new_data |= ret;
+
+	/* New information is found, we need to save to the cache */
+	if (has_new_data)
+		kerndat_save_cache();
```

Comment:
Why do we need the `has_new_data` variable? Could we do something like the following instead?

```C
int kerndat_try_load_new(void)
{
	int ret;

	ret = kerndat_get_hugetlb_dev();
	if (ret < 0)
		return ret;

	/* New information is found, we need to save to the cache */
	if (!ret)
		kerndat_save_cache();

	return 0;
}
```

Reply:
At first, I want to add a kerndat check for mremap with hugetlb in this function too. However, as my machine kernel does not have that feature, I need to build a kernel for testing it so maybe it will be in later pull request.